src/protocol/server.c

```diff
@@ -198,7 +253,166 @@ protocolServerProcess(
+                                            }
+                                        }
+
+                                        // Process request type
+                                        switch (request.type)
+                                        {
+                                            // Process request
+                                            case protocolCommandTypeProcess:
+                                            {
+                                                // Process session
+                                                ProtocolServerResult *processResult;
+
+                                                if (handler->processSession != NULL)
+                                                {
+                                                    ASSERT(handler->process == NULL);
+                                                    CHECK_FMT(
+                                                        ProtocolError, this->sessionId != 0, "no session id for request %s:%s",
+                                                        strZ(strIdToStr(request.id)), strZ(strIdToStr(request.type)));
+
+                                                    processResult = handler->processSession(pckReadNew(request.param), sessionData);
+                                                }
+                                                // Standalone process
+                                                else
+                                                {
+                                                    ASSERT(handler->process != NULL);
+                                                    ASSERT(!request.sessionRequired);
+
+                                                    processResult = handler->process(pckReadNew(request.param));
+                                                }
+
+                                                if (processResult == NULL)
+                                                    protocolServerResponseP(this);
+                                                else
+                                                {
+                                                    ASSERT(processResult->sessionData == NULL);
+                                                    ASSERT(handler->processSession != NULL || !processResult->close);
+
+                                                    protocolServerResponseP(
+                                                        this, .data = processResult->data, .close = processResult->close);
+
+                                                    // Free session when close is true. This optimization allows an explicit
+                                                    // close to be skipped.
+                                                    if (processResult->close)
+                                                    {
+                                                        objFree(sessionData);
```

Comment:
While this lets us avoid an explicit close round-trip, it seems that the `case protocolCommandTypeClose` below will call the handler's close routine; why do we not also need to potentially handle this here?

Reply:
Because the handler has already done the closing action if any is needed. In fact, the handler requesting close and having an explicit close request defined are mutually exclusive. They don't need to be but in practice they are (for now). Explicit close is only used by PROTOCOL_COMMAND_STORAGE_WRITE and the close request by the handler is only used by PROTOCOL_COMMAND_STORAGE_READ. But that's not required, just the way it is now.